File: tiq-e/linked-list/3/main.ts
import {performance} from "perf_hooks";

class ListNode {
    val: number;
    next: ListNode | null;

    constructor(val?: number, next?: ListNode | null) {
        this.val = val === undefined ? 0 : val;
        this.next = next === undefined ? null : next;
    }
}

function reverseList(head: ListNode | null): ListNode | null {
    let pre = null;
    while (head) {
        const next = head.next;
        head.next = pre;
        pre = head;
        head = next;
    }
    return pre;
}

const head = new ListNode(1, null);

const start = performance.now();
const result = reverseList(head);
const end = performance.now();
console.log("Duration:", end - start);
console.log("Result:", result);

File: tiq-e/linked-list/3/main.py
from collections import deque
from time import perf_counter
from typing import List


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        t = None
        d = ListNode(-1)

        while head:
            th1 = th2 = head
            head = head.next

            th1.next = t
            t = th2

            if not head:
                d.next = th2
                break

        return d.next


head = ListNode(1, None)

start = perf_counter()
result = Solution().reverseList(head)
end = perf_counter()
print("Duration: " + str(end - start))
print("Result: " + str(result))
